rfcs/77-dev-tools-abstraction.md

```diff
@@ -0,0 +1,634 @@
+    }
+
+    /// Gets a mutable reference to the specified modal dev tool by name.
+    fn get_tool_mut_by_name(&self, mut world: &mut World, component_id: ComponentId) -> Option<&mut dyn ModalDevTool> {
+        let component_id = self.lookup_tool_component_id(name)?;
+        DevToolsRegistry::get_mut_by_id(world, component_id)
+    }
+    
+    /// Iterates over the list of registered modal dev tools.
+    fn iter_tools(&self, world: &World) -> impl Iterator<Item = (ComponentId, &dyn ModalDevTool)> { 
+        self.modal_dev_tools.iter().map(|&id| (id, self.get(world, id)))
+    }
+
+    /// Mutably iterates over the list of registered modal dev tools.
+    fn iter_tools_mut(&self, mut world: &mut World) -> impl Iterator<Item = (ComponentId, &mut dyn ModalDevTool)> { 
+        self.modal_dev_tools.iter_mut().map(|&id| (id, self.get(world, id)))
+    }
+
+    /// Iterates over the list registered dev commands, returning their name and `DevCommandMetadata`.
+    fn iter_commands(&self) -> impl Iterator<Item = (&str, `DevCommandMetadata)> {
+        self.dev_commands.iter()
+    }
+}
+```
+
+Once the user has a `dyn ModalDevTool`, they can perform whatever operations they'd like: enabling and disabling it, reading metadata and so on.
+There's no need to add further duplicate API: users building toolboxes are generally sophisticated, and the number of calls should be quite small.
+
+While we can use the [dynamic resource APIs](https://docs.rs/bevy/latest/bevy/ecs/world/struct.World.html#method.get_resource_by_id) and trait casting to get references to the list of modal dev tools, we cannot do the same for the dev commands!
+While this may seem unintuitive, the reason for this is fairly simple: modal configuration is always present, while the configuration for each dev command is generated when it is sent.
+As a result, we can only access its metadata: the arguments it takes, its doc strings and so on.
+
+### Parsing and inserting modal dev tools
+
+One method on `DevToolsRegistry` is worth special attention: `parse_and_insert_tool`.
+This tool takes a common pattern, parsing the configuration for a dev tool from a provided string, and provides a user friendly, safe API over it.
+
+```rust
+impl DevToolsRegistry {
+    /// Parses the given str `s` into a modal dev tool corresponding to its name if possible.
+    ///
+    /// For a typed equivalent, simply use the `FromStr` trait that this method relies on directly.
+    fn parse_and_insert_tool(&self, world: &mut World, s: &str) -> Result<(), DevToolParseError>{
+        // Parse out the name
+        let name = s.clone().split_whitespace.next()?;
+        
+        // Get the associated `ComponentId`, so we can use it to insert a resource of a dynamic type
+        let component_id = self.lookup_tool_component_id(name);
+        // Look-up the existing resource to get access to get access to the metadata we need
+        let tool_metadata: &dyn DevTool = self.get_tool_metadata(component_d)?;   
```

Comment:
```suggestion
        let tool_metadata: &dyn DevTool = self.get_tool_metadata(component_id)?;   
```

Reply:
Lets say we return Metadata here, instead of &dyn DevTool. If we go with the enum approach, we can match arms and one or another, depending on the tool type.